truenas_verify/mtree_verify.py

```diff
@@ -0,0 +1,128 @@
+
+
+def parse_mtree_entry(line):
+    if line.startswith('#'):
+        return None
+
+    fname, mode, gid, uid, extra = line[1:].split(maxsplit=4)
+    if extra.startswith('type=dir'):
+        entry = MTREE_ENTRY(
+            fname,
+            mode.split('=')[1],
+            int(uid.split('=')[1]),
+            int(gid.split('=')[1]),
+            'dir',
+            None,
+            None,
+            None
+        )
+    elif extra.startswith('type=link'):
+        ftype, link = extra.split()
+        entry = MTREE_ENTRY(
+            fname,
+            mode.split('=')[1],
+            int(uid.split('=')[1]),
+            int(gid.split('=')[1]),
+            'link',
+            link.split('=')[1],
+            None,
+            None
+        )
+    else:
+        ftype, size, shasum = extra.split()
+        entry = MTREE_ENTRY(
+            fname,
+            mode.split('=')[1],
+            int(uid.split('=')[1]),
+            int(gid.split('=')[1]),
+            ftype.split('=')[1],
+            None,
+            int(size.split('=')[1]),
+            shasum.split('=')[1]
+        )
+
+    return entry
+
+
+def validate_file_sha256sum(entry, errors):
+    with open(entry.fname, 'rb', buffering=0) as f:
+        hash = file_digest(f, 'sha256').hexdigest()
+        if hash != entry.sha256:
+            errors.append(f'{entry.fname}: expected: {entry.sha256}, got: {hash}')
+
+
+def validate_mtree_entry(entry) -> list[str]:
+    try:
+        st = lstat(entry.fname)
+    except FileNotFoundError:
+        return [f'{entry.fname}: file does not exist.']
+
+    assert st.st_uid == entry.uid, entry.fname
+    assert st.st_gid == entry.gid, entry.fname
+
+    errors = []
+    match entry.type:
+        case 'dir':
+            if not S_ISDIR(st.st_mode):
+                errors.append(f'{entry.fname}: incorrect file type.')
+        case 'file':
+            if not S_ISREG(st.st_mode):
+                errors.append(f'{entry.fname}: incorrect file type.')
+
+            validate_file_sha256sum(entry, errors)
+        case 'link':
+            if not S_ISLNK(st.st_mode):
+                errors.append(f'{entry.fname}: incorrect file type.')
+
+    if oct(S_IMODE(st.st_mode))[2:] != entry.mode:
+        errors.append(f'{entry.fname}: got mode {oct(S_IMODE(st.st_mode))}, expected: {entry.mode}')
+
+    return errors
+
+
+def load_mtree_file():
+    with open(MTREE_FILE_PATH, 'r') as f:
+        chunk_size = 1000
+        chunk = []
+        for line in f:
+            chunk.append(line)
+            if len(chunk) >= chunk_size:
+                yield ''.join(chunk)
+                chunk.clear()
+        if chunk:
+            yield ''.join(chunk)
+
+
+def process_chunk(chunk) -> list[str]:
+    errors = []
+    for line in chunk.splitlines():
+        if (entry := parse_mtree_entry(line)) is not None:
+            errors.extend(validate_mtree_entry(entry))
+    return errors
+
+
+def main():
+    with Pool() as pool:
```

Comment:
> processes is the number of worker processes to use. If processes is None then the number returned by [os.cpu_count()](https://docs.python.org/3/library/os.html#os.cpu_count) is used.

So we should probably set an upper-bound on processes for the process pool.

Reply:
Well, on one of my system that'd be 32 which is somewhat excessive :)

Reply:
Point taken